src/registrar/models/domain.py

```diff
@@ -0,0 +1,223 @@
+import logging
+import re
+
+from django.db import models
+from django_fsm import FSMField, transition  # type: ignore
+
+from epp.mock_epp import domain_info, domain_check
+
+from .utility.time_stamped_model import TimeStampedModel
+from .domain_application import DomainApplication
+from .user import User
+
+logger = logging.getLogger(__name__)
+
+
+class Domain(TimeStampedModel):
+    """
+    Manage the lifecycle of domain names.
+
+    The registry is the source of truth for this data and this model exists:
+        1. To tie ownership information in the registrar to
+           DNS entries in the registry; and
+        2. To allow a new registrant to draft DNS entries before their
+           application is approved
+    """
+
+    class Meta:
+        constraints = [
+            # draft domains may share the same name, but
+            # once approved, they must be globally unique
+            models.UniqueConstraint(
+                fields=["name"],
+                condition=models.Q(is_active=True),
+                name="unique_domain_name_in_registry",
+            ),
+        ]
+
+    class Status(models.TextChoices):
+        """The status codes we can receive from the registry."""
```

Comment:
Can't add an exact link because they are buried deep in a long document, but I've added a reference and a link to the document. (Could probably find a linkable version of that RFC, but it makes more sense I think to have the whole entire thing in context, if someone were to go hunting for more information.)